src/host/services/IMUService.cpp

```diff
@@ -26,38 +26,191 @@
 #include "common/signalk/SKUpdateStatic.h"
 #include "common/signalk/SKUnits.h"
 #include "IMUService.h"
+#include "host/util/PersistentStorage.h"
 
 void IMUService::setup() {
+  _timeSinceLastSetOffset = 0;
+  _offsetRoll  = 0.0;
+  _offsetPitch = 0.0;
+  // Default mounting position: VerticalStbHull
+  uint8_t axisConfig = 0b00001001;
+  uint8_t signConfig = 0b00000000;
+
+  if (_config.mounting == VerticalTopToBow) {
+    axisConfig = 0b00001001;
+    signConfig = 0b00000000;
+  }
+  if (_config.mounting == HorizontalLeftSideToBow) {
+    axisConfig = 0b00100100;
+    signConfig = 0b00000000;
+  }
+
   DEBUG("Initing BNO055");
-  if (!bno055.begin()) {
+  if (!bno055.begin(bno055.OPERATION_MODE_NDOF, axisConfig, signConfig)) {
     DEBUG("Error initializing BNO055");
   }
   else {
-    DEBUG("Success!");
+    // read calibration and offset values for heel & pitch from flash
+    readCalibrationOffsets();
+    readHeelPitchOffset();
   }
 }
 
 void IMUService::loop() {
-  bno055.getCalibration(&sysCalib, &gyroCalib, &accelCalib, &magCalib);
+  bno055.getCalibration(&_sysCalib, &_gyroCalib, &_accelCalib, &_magCalib);
+  //DEBUG("Calib Sys: %i Accel: %i Gyro: %i Mag: %i", _sysCalib, _accelCalib, _gyroCalib, _magCalib);
 
   eulerAngles = bno055.getVector(Adafruit_BNO055::VECTOR_EULER);
-
-  //DEBUG("Calib Sys: %i Accel: %i Gyro: %i Mag: %i", sysCalib, accelCalib, gyroCalib, magCalib);
-  //DEBUG("Attitude roll: %f pitch: %f  Mag heading: %f", eulerAngles.z(), eulerAngles.y(), eulerAngles.x());
+  //DEBUG("Sensor Raw-datas: eulerAngles.z=%.3f eulerAngles.y=%.3f  eulerAngles.x=%.3f", eulerAngles.z(), eulerAngles.y(), eulerAngles.x());
 
   SKUpdateStatic<2> update;
   // Note: We could calculate yaw as the difference between the Magnetic
   // Heading and the Course Over Ground Magnetic.
 
-  /* if orientation == MOUNTED_ON_PORT_HULL */
-  double roll, pitch, heading;
-  roll = eulerAngles.z();
-  pitch = eulerAngles.y();
-  heading = fmod(eulerAngles.x() + 270, 360);
+  if (_config.mounting == VerticalStbHull) {
```

Comment:
We should use a switch statement here. The compiler will throw a warning for the values of the enum that we are not visiting.

Reply:
I think this config has a slight issue: The pitch up is measured as negative. It should be positive.

Reply:
> We should use a switch statement here. The compiler will throw a warning for the values of the enum that we are not visiting.
Done

> I think this config has a slight issue: The pitch up is measured as negative. It should be positive

I took your original settings, but I think I mixed up positions. Now port hull is default.
But all positions should be tested, if they are correct, may be you could do it, or I will do, when I am back home next week.